apps/src/code-studio/pd/form_components_func/FormController.jsx

```diff
@@ -328,41 +358,25 @@ const FormController = props => {
     setGlobalError(false);
     setSubmitting(true);
 
-    // [MEG] TODO: only do a post if it's a new application
-    $.ajax({
-      method: 'POST',
-      url: apiEndpoint,
-      contentType: 'application/json',
-      dataType: 'json',
-      data: JSON.stringify(serializeFormData())
-    })
-      .done(data => {
-        sessionStorage.removeItem(sessionStorageKey);
-        onSuccessfulSubmit(data);
-      })
-      .fail(data => {
-        if (
-          data.responseJSON &&
-          data.responseJSON.errors &&
-          data.responseJSON.errors.form_data
-        ) {
-          if (data.responseJSON.general_error) {
-            setErrors(data.responseJSON.errors.form_data);
-            setErrorHeader(data.responseJSON.general_error);
-            setGlobalError(true);
-          } else {
-            // if the failure was a result of an invalid form, highlight the errors
-            // and display the generic error header
-            setErrors(data.responseJSON.errors.form_data);
-            setErrorHeader(i18n.formErrorsBelow());
-          }
-        } else {
-          // Otherwise, something unknown went wrong on the server
-          setGlobalError(true);
-          setErrorHeader(i18n.formServerError());
-        }
-        setSubmitting(false);
-      });
+    const handleSuccessfulSubmit = data => {
+      sessionStorage.removeItem(sessionStorageKey);
+      onSuccessfulSubmit(data);
+    };
+
+    const handleRequestFailure = data => {
+      if (get(data, 'responseJSON.errors.form_data')) {
+        setGlobalError(true);
+        setErrors(data.responseJSON.errors.form_data);
+        setErrorHeader(i18n.formErrorsBelow());
+      } else {
+        // Otherwise, something unknown went wrong on the server
+        setGlobalError(true);
+        setErrorHeader(i18n.formServerError());
```

Comment:
handleRequestFailure is different functionality. I removed the `if (data.responseJSON.general_error)` block because I didn't see evidence it ever got there. I also added `setGlobalError(true)` when there were `errors.form_data`. @tim-dot-org does that look okay to you?

Reply:
i think we don't want the extra setGlobalError, if i understand it correctly, later on we have: ```const shouldShowError = errorHeader && (globalError || pageHasError());```
and we might be on a page that doesn't have errors, but the error header is set and we don't want to show the error pop up. If the page has page specific errors, we handle that using pageHasError.

Reply:
Ah good call. Sounds good––will remove it (and add a test to avoid confusion)

Reply:
Confusingly, we set globalError on `if (invalidPages.length) {` too ...

Reply:
This is the scenario I'm confused about: I hit the submit button, and handleSubmit is called. We first make sure the required fields have what they need. Only then will we make the request to the server. If the request fails with form_data errors, we `setError` and `setErrorHeader`. Suppose pageHasErrors() isn't true ... then we'd just scroll to the top and no errors would be shown ... I feel like that would be kinda bizarre (scroll to the top with no errors)

Reply:
I'm not entirely sure what the "right" approach is ... but currently, if we are somehow getting validation errors on the server that the client isn't catching, I think we're setting up the user for confusion

Reply:
Okay, I just kept it as-is for now (no setGlobalError when there are form errors as before, keep setPageInternal the same). I'm not confident enough about a refactor to warrant the change